### ISSUE TITLE
[Feature request] Japanese language

### ISSUE THREAD

**Issue by jonathan-s** [futo-org/android-keyboard#42]
Adding my vote for japanese language which I'm currently missing.

**Comment by nonetrix** [comment]
As of now the only open source keyboard that supports Japanese is some [weird Google Mozc app](https://f-droid.org/en/packages/org.mozc.android.inputmethod.japanese/) that looks really dated and isn't ideal, this would make me be able to switch from gboard without internet access completely

**Comment by vdmkenny** [comment]
The 12key input would be greatly appreciated. There is no modern privacy oriented keyboard that supports it.

**Comment by Claude-Kenni** [comment]
Yeah, Japanese flick keyboard

**Comment by jedinjapan** [comment]
In addition to 12-Key, please include QWERTY layout and Swipe matching for this language as well. This was a feature of the Swype keyboard almost 6-7 years ago I have not seen replicated elsewhere.

If you are used to typing kana on 12-Key, that is pretty much the standard in Japan. But for many second language learners, and for some Japanese typists, they never master this. But the QWERTY input method can allow matching against the autocomplete dictionary for romaji input, akin to what is done in Windows with the Microsoft IME or ATOK.

In theory, this would be easier to do using the existing QWERTY layout and the swipe implementations already implemented. This could additionally serve as a "stop-gap" while the 12-key layout is developed.

**Comment by Edoardo-Morosanu** [comment]
I've forked the repository today and plan to at least start working on it over the next few weeks, unless someone else is already actively developing it. The only viable way I found so far to implement that would be by integrating [mozc](https://github.com/google/mozc), which seems to also be what GBoard uses. A working implementation of this method would be [MozcForAndroid](https://github.com/kachaya/MozcForAndroid).

**Comment by FrostWillow** [comment]
Chiming in this request for the Japanese 12 key flick keyboard please. It's something I use daily and is the only thing keeping me from using Futo Keyboard exclusively. Even if we only get the QWERTY version at first that's a fine first step and would make waiting for a 12 key flick keyboard easier, just something that allows me to write in Japanese without switching to a different keyboard entirely would be great.

Thanks for all your hard work on this keyboard so far.

**Comment by TheQwervy** [comment]
Adding my vote for Japanese support, flick or qwerty

**Comment by nonetrix** [comment]
Why not both?

**Comment by FrostWillow** [comment]
> Why not both? There is also this which some use on tablets <img alt="image" width="602" src="https://private-user-images.githubusercontent.com/45698918/357129756-73d1001b-295f-4ff5-8b2f-af86a477949f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM5MDY0MzgsIm5iZiI6MTcyMzkwNjEzOCwicGF0aCI6Ii80NTY5ODkxOC8zNTcxMjk3NTYtNzNkMTAwMWItMjk1Zi00ZmY1LThiMmYtYWY4NmE0Nzc5NDlmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE3VDE0NDg1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZjMzJhNWM4ZTUxZmYwN2U5YTk3ZjZhN2ViNGY0YmIyMDdhNTJhZTZmYjY1MzdkNjk2MTI4ZTdjMmVlOWQxY2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.sLLgj4v3aUcc8gw4PRF36m0Wwxy1M1vEEesx45W-5hg">

Oh please no, That would be a terrible option. Even Japanese people don't like using that keyboard. They mostly only use either the standard QWERTY, which is usually used for desktops, or the flick keyboard. 

I'm sure there are people out there who use that and like it, but that's like the Japanese equivalent of the Dvorak keyboard. Niche and only really liked by the few people who got really into using it.

**Comment by mvevitsis** [comment]
> Why not both? There is also this which some use on tablets
> <img width="602" alt="image" src="https://github.com/user-attachments/assets/73d1001b-295f-4ff5-8b2f-af86a477949f">
> 

No one uses that. And pretty much no one uses romaji input on their phone either. Flick or die.

**Comment by mvevitsis** [comment]
> It was just a suggestion, why not have all 3. I don't see the problem, it is common enough to be included in iOS pretty sure

More is better, sure. But I would prefer effort be spent on the most used option first, especially since they can copy/paste most of the code from mozc.

**Comment by jedinjapan** [comment]
The fact is there are 3 possible Japanese keyboard layouts, with fairly universal agreement on which are most popular. From most popular to least:

1. 12-Key Input - This is a "numpad" style that use to be tap-based but now uses directional flick inputs. For English speakers and devs unfamiliar with it this is similar to T9 texting, but easier due to how Japanese uses vowel/consonant pairs in kana. Mozc repo features this. Extremely popular input method. Present in Google Japanese Input and Microsoft Swiftkey.

2. QWERTY Keyboard - aka "romaji" input. This is used commonly on Japanese PCs, and translates 1:1 to virtual keyboards. Additionally some virtual keyboards offer swipe functionality for this, but it is exceedingly rare. Highly desireable if possible. Use would be lower however, used by many non-native speakers and perhaps those who never learned 12-key. It would however be an "easy to implement" option in theory if the roman text to kana conversion is easy to copy over and setup, and a good "holdover" option if this is the case. Mozc also features this. Low hanging fruit, somewhat niche, but useful still. Microsoft Swiftkey and Google Japanese Input both feature an implementation of this without Swipe functionality. Swype keyboard (discontinued) featured swiping.

3. Kana Keyboard - As show in the image above, this is available on Japanese laptops, but even domestically in Japan this is almost never used, and is largely a holdover from the 1980s and 1990s Japanese tech industry. Extremely rare to see in a virtual keyboard, almost never used in a physical keyboard when present. If implemented, will be used by less than 1% of users without a doubt. Still a neat curiosity, and ironically probably one of the most straight forward to implement as it's just directly typing the kana symbols.

All 3 above options would need the same Kana input to Kanji conversion function, so they share difficult on that. This code is in Mozc as well I believe.

The Mozc project will offer a strong starting point for options 1 and 2 potentially, though a transformer dictionary would need to be trained for Futo Keyboard. Mozc has a 12 key format built in that can be forked, and the existing Futo QWERTY can be used for option 2 if implemented. The Mozc license appears to be extremely permissive, but I leave that to devs to review.

Mozc (aka Google Japanese Input) Github:
https://github.com/google/mozc

For further context: Google Japanese Input is really excellent, and used by companies and users within Japan fairly frequently in lieu of the sometime awful software keyboards the domestic cell phone companies implemented, and the stock android keyboards. Also has taken a huge chunk out of IME company ATOK's market share on desktop by virtue of being 99% as good, and free for corporations.

**Comment by nonetrix** [comment]
> > It was just a suggestion, why not have all 3. I don't see the problem, it is common enough to be included in iOS pretty sure
> 
> More is better, sure. But I would prefer effort be spent on the most used option first, especially since they can copy/paste most of the code from mozc.
> 
> The others are technically less complicated to do, but there is also less open source code to use as a reference.
> 
> For romaji input -> fcitx5 has this For kana (not 12 key) input -> florisboard sort of has it implemented

It isn't rocket science to add I don't see reason not to have it, like I could probably add it myself if QWERTY is added and I don't have much experience

**Comment by jedinjapan** [comment]
> I don't see reason not to have it

Long-term maintenance of the feature could cause unforeseen bugs and complications. It will be used by <1% of possible users, if that. This would be an academic curiosity. If the dev finds it fun to implement in downtime as a laugh then sure, but otherwise it's a development dead-end in many ways, and does not come with 0 cost to other features.

**Comment by TheQwervy** [comment]
> Edit: put another way, if the direct Kana keyboard layout is implemented as an option, my very next request would be the option to fully disable it, so I do not accidentally switch to it in rotation when moving between QWERTY/Romaji and 12-Key inputs.

The way that keyboards are added currently you wouldn't have to worry about that because you chose the layout and language per option.

**Comment by jonathan-s** [comment]
> > Why not both? There is also this which some use on tablets

> No one uses that. And pretty much no one uses romaji input on their phone either. Flick or die.

I use the romaji option all the time, and so does my japanese partner.

**Comment by ScoptOriginal** [comment]
Yeah, I'm absolutely desperate for a Japanese 12-key layout, it's the one thing holding me back from switching to FUTO full time
